### ISSUE TITLE
Performance issue using multiple ComputationalGraph outputs

### ISSUE THREAD

**Issue by snurkabill** [deeplearning4j/deeplearning4j#6544]
Hello there, 

having performance issues with training and evaluation of multidataset on ComputationGraph. I've created gist for the better explanation problem of mine. 

It is very possible that I am doing something wrong, but I haven't found some nice how-to for multidataset iterators. 

Here is gist:
https://gist.github.com/snurkabill/bd47abfb1a15fcfcdb73e82ccb647d3b

there are three copy-pasted files with minor changes wrapping DL4J models.  Minor changes are outputs. In the first (preferred) model, there are three output nodes, each one with three different activations and loss functions. Second model with only two and the third one with only one output node. 

Each output node takes as input the same hidden layer(graph node), so the parameter count is still the same but training and evaluation performance is quite different. 

Reproducer file creates a quite small dataset as a demonstration and then trains the model on it.  (there are commented models 1, 2, 3... one can just switch them with commenting the other one)

Training between first and third model differs by a lot. Thrid model takes only 700ms on my machine, but the first one takes 2300ms! that is quite a difference since the count of parameters is the same in the network topology. 

Measurement is not perfect and precise, this is not benchmark but benchmarking is not needed. It is visibly slower. One can wait for few training iterations for proper JVM warmup. 

This is log on my machine when creating model: 
17:38:28.942 [main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
17:38:29.996 [main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 4
17:38:30.157 [main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 4
17:38:30.157 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Windows 8.1]
17:38:30.157 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [8]; Memory: [4.0GB];
17:38:30.157 [main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [MKL]
17:38:30.302 [main] INFO org.deeplearning4j.nn.graph.ComputationGraph - Starting ComputationGraph with WorkspaceModes set to [training: ENABLED; inference: ENABLED], cacheMode set to [NONE]

Interestingly, only one CPU core seems to be working while training/evaluating, which doesn't correspond to the number of thread used in log. 

So my question is, what am I doing wrong, please. I really need to get this fast as possible and I've chosen DL4J because java is love and DL4J supposed to be quite fast. It doesn't seem like it, but I really think I am doing something terrible with iterators. 

Thank you for help.

**Comment by AlexDBlack** [comment]
Somehow I mised this one until now.

I ran your code locally. You're doing 1000 epochs, 4000 iterations. So 700-2000ms is not exactly slow, given what you are doing (that's 0.175ms/iteration to 0.5ms/iteration).


A few observations:
1. Your network is really tiny. For many operations, it'll be faster to use a single thread for parallelizing operations - which might explain why you're seeing a single thread being used.
2. There are fixed overheads (for native method calls amongst other things) - it's not simply linear in the number of parameters
3. Have a look at our benchmarking guide: https://deeplearning4j.org/docs/latest/deeplearning4j-benchmark
